lightning/src/chain/onchain_utils.rs

```diff
@@ -0,0 +1,884 @@
+			&InputDescriptors::RevokedOutput => {
+				writer.write_all(&[4; 1])?;
+			}
+		}
+		Ok(())
+	}
+}
+
+impl Readable for InputDescriptors {
+	fn read<R: ::std::io::Read>(reader: &mut R) -> Result<Self, DecodeError> {
+		let input_descriptor = match <u8 as Readable>::read(reader)? {
+			0 => {
+				InputDescriptors::RevokedOfferedHTLC
+			},
+			1 => {
+				InputDescriptors::RevokedReceivedHTLC
+			},
+			2 => {
+				InputDescriptors::OfferedHTLC
+			},
+			3 => {
+				InputDescriptors::ReceivedHTLC
+			},
+			4 => {
+				InputDescriptors::RevokedOutput
+			}
+			_ => return Err(DecodeError::InvalidValue),
+		};
+		Ok(input_descriptor)
+	}
+}
+
+/// Sum up witnesses weight of a set of given inputs. Inputs might belong to differing descriptors.
+/// Weights should be compliant with BOLT 3 Appendix A.
+pub(crate) fn get_witnesses_weight(inputs: &[InputDescriptors]) -> usize {
+	let mut tx_weight = 2; // count segwit flags
+	for inp in inputs {
+		// We use expected weight (and not actual) as signatures and time lock delays may vary
+		tx_weight +=  match inp {
+			// number_of_witness_elements + sig_length + revocation_sig + pubkey_length + revocationpubkey + witness_script_length + witness_script
+			&InputDescriptors::RevokedOfferedHTLC => {
+				1 + 1 + 73 + 1 + 33 + 1 + 133
+			},
+			// number_of_witness_elements + sig_length + revocation_sig + pubkey_length + revocationpubkey + witness_script_length + witness_script
+			&InputDescriptors::RevokedReceivedHTLC => {
+				1 + 1 + 73 + 1 + 33 + 1 + 139
+			},
+			// number_of_witness_elements + sig_length + counterpartyhtlc_sig  + preimage_length + preimage + witness_script_length + witness_script
+			&InputDescriptors::OfferedHTLC => {
+				1 + 1 + 73 + 1 + 32 + 1 + 133
+			},
+			// number_of_witness_elements + sig_length + revocation_sig + pubkey_length + revocationpubkey + witness_script_length + witness_script
+			&InputDescriptors::ReceivedHTLC => {
+				1 + 1 + 73 + 1 + 1 + 1 + 139
+			},
+			// number_of_witness_elements + sig_length + revocation_sig + true_length + op_true + witness_script_length + witness_script
+			&InputDescriptors::RevokedOutput => {
+				1 + 1 + 73 + 1 + 1 + 1 + 77
+			},
+		};
+	}
+	tx_weight
+}
+
+/// A struct to describe a revoked output and corresponding informations to generate a solving
+/// witness.
+///
+/// If the revoked output is a HTLC output on a committment transaction, HTLCOuputInCommitment (hash
+/// timelock, direction), pubkeys are used to generate the suiting witnessScript.
+///
+/// Otherwise, CSV is used to generate a witnessScript redeeming a balance output on a commitment
+/// transaction or second-stage HTLC transaction.
+///
+/// Amount and revocation secret are used in both cases.
+#[derive(Clone, PartialEq)]
+pub(crate) struct RevokedOutput {
+	per_commitment_point: PublicKey,
+	counterparty_delayed_payment_base_key: PublicKey,
+	counterparty_htlc_base_key: PublicKey,
+	per_commitment_key: SecretKey,
+	input_descriptor: InputDescriptors,
```

Comment:
Its a little confusing that a `RevokedOutput` could have type `OfferedHTLC`/`ReceivedHTLC`. Can we split out the type here or assert that the type is a `Revoked`... one in `build` (it looks like the non-revoked types are really just used for `get_witnesses_weight`).

Reply:
So I dried-up InputDescriptors and instead I'm using new named constants. It turns out InputDescriptors and the associated helper `get_witnesses_weight` don't serve a real purpose. At detection, spend weights can be marked as static, we won't touch them anymore until transaction generation.